presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayReduceFunction.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.operator.scalar;
+
+import com.facebook.presto.metadata.BoundVariables;
+import com.facebook.presto.metadata.FunctionKind;
+import com.facebook.presto.metadata.FunctionRegistry;
+import com.facebook.presto.metadata.Signature;
+import com.facebook.presto.metadata.SqlScalarFunction;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.TypeManager;
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableList;
+import com.google.common.primitives.Primitives;
+
+import java.lang.invoke.MethodHandle;
+
+import static com.facebook.presto.metadata.Signature.typeVariable;
+import static com.facebook.presto.spi.type.TypeSignature.parseTypeSignature;
+import static com.facebook.presto.spi.type.TypeUtils.readNativeValue;
+import static com.facebook.presto.util.Reflection.methodHandle;
+
+public final class ArrayReduceFunction
+        extends SqlScalarFunction
+{
+    public static final ArrayReduceFunction ARRAY_REDUCE_FUNCTION = new ArrayReduceFunction();
+
+    private static final MethodHandle METHOD_HANDLE = methodHandle(ArrayReduceFunction.class, "reduce", Type.class, Type.class, Type.class, Block.class, Object.class, MethodHandle.class, MethodHandle.class);
+
+    private ArrayReduceFunction()
+    {
+        super(new Signature(
+                "reduce",
+                FunctionKind.SCALAR,
+                ImmutableList.of(typeVariable("T"), typeVariable("S"), typeVariable("R")),
+                ImmutableList.of(),
+                parseTypeSignature("R"),
+                ImmutableList.of(parseTypeSignature("array(T)"), parseTypeSignature("S"), parseTypeSignature("function(T,S,S)"), parseTypeSignature("function(S,R)")),
```

Comment:
Maybe we should also have a form that's `reduce(array(T), R, function(T, R, R))`? Since it's probably common to have an identity transformation from the intermediate state to the final state